server/src/main/java/org/elasticsearch/action/admin/cluster/node/reload/NodesReloadSecureSettingsRequest.java

```diff
@@ -19,22 +19,96 @@
 
 package org.elasticsearch.action.admin.cluster.node.reload;
 
+import org.elasticsearch.Version;
 import org.elasticsearch.action.support.nodes.BaseNodesRequest;
+import org.elasticsearch.common.CharArrays;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.bytes.BytesArray;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.settings.SecureString;
+
+import java.io.IOException;
+import java.util.Arrays;
 
 /**
- * Request for a reload secure settings action.
+ * Request for a reload secure settings action
  */
 public class NodesReloadSecureSettingsRequest extends BaseNodesRequest<NodesReloadSecureSettingsRequest> {
 
+    /**
+     * The password is used to re-read and decrypt the contents
+     * of the node's keystore (backing the implementation of
+     * {@code SecureSettings}).
+     */
+    @Nullable
+    private SecureString secureSettingsPassword;
+
     public NodesReloadSecureSettingsRequest() {
     }
 
     /**
-     * Reload secure settings only on certain nodes, based on the nodes IDs specified. If none are passed, secure settings will be reloaded
-     * on all the nodes.
+     * Reload secure settings only on certain nodes, based on the nodes ids
+     * specified. If none are passed, secure settings will be reloaded on all the
+     * nodes.
      */
-    public NodesReloadSecureSettingsRequest(final String... nodesIds) {
+    public NodesReloadSecureSettingsRequest(String... nodesIds) {
         super(nodesIds);
     }
 
+    @Nullable
+    public SecureString getSecureSettingsPassword() {
+        return secureSettingsPassword;
+    }
+
+    public NodesReloadSecureSettingsRequest setSecureStorePassword(SecureString secureStorePassword) {
+        this.secureSettingsPassword = secureStorePassword;
+        return this;
+    }
+
+    public void closePassword() {
+        if (this.secureSettingsPassword != null) {
+            this.secureSettingsPassword.close();
+        }
+    }
+
+    boolean hasPassword() {
+        return this.secureSettingsPassword != null && this.secureSettingsPassword.length() > 0;
+    }
+
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
+        super.readFrom(in);
+        if (in.getVersion().onOrAfter(Version.V_8_0_0)) {
+            final BytesReference bytesRef = in.readOptionalBytesReference();
+            if (bytesRef != null) {
+                byte[] bytes = BytesReference.toBytes(bytesRef);
```

Comment:
A somewhat orthogonal point, but it seems like most of the places we use CharArrays like this we have a BytesReference, and we end up copying to a byte array in order to convert to chars. I wonder if we could flip the CharArrays input around so it takes a BytesReference, so that we don't need to have the try/finally like this. In the few cases we actually have a real bytes array, we can wrap the array in a bytesreference. It's not necessary to do here, but I think it would make these uses much easier to deal with and not so messy.

Reply:
I'll make an issue out of your comment and tackle it in all cases we have this in a follow up PR